numpyro/distributions/continuous.py

```diff
@@ -290,6 +291,105 @@ def infer_shapes(concentration):
         return batch_shape, event_shape
 
 
+class EulerMaruyama(Distribution):
+    arg_constraints = {"t": constraints.ordered_vector}
+
+    def __init__(self, t, sde_fn, sde_pars, init_dist, validate_args=None):
+        # we use t rather than (dt, num_steps) to make the api simpler and more general,
+        # `init_dist` will define prior for the init value,
+        # event dimensions will be decided by `init_dist`
+        self.t = t
+        self.dt = jnp.diff(t)
+        self.sde_fn = sde_fn
+        self.sde_pars = sde_pars
+        self.init_dist = init_dist
+
+        if init_dist.batch_shape:
+            raise ValueError(
+                "Init distribution is expected to have no batch dimension."
+            )
+        if jnp.ndim(t) != 1:
+            raise ValueError("`t` is expected to have 1 dimension.")
+        event_shape = t.shape + init_dist.event_shape
+        batch_shape = ()
+
+        super(EulerMaruyama, self).__init__(
+            batch_shape, event_shape, validate_args=validate_args
+        )
+
+    @constraints.dependent_property
+    def support(self):
+        return constraints.independent(constraints.real, self.event_dim)
+
+    def sample(self, key, sample_shape=()):
+        assert is_prng_key(key)
+
+        sample_size = np.prod(sample_shape)
+        _sample_shape = (sample_size,) if sample_shape else ()
+
+        def step(y_curr, xs):
+            t_curr, dt_curr, noise_curr = xs
+            sde_pars = [
+                x.sample(key, sample_shape=_sample_shape) for x in self.sde_pars
+            ]
+            f, g = self.sde_fn(y_curr, t_curr, *sde_pars)
+            mu = y_curr + dt_curr * f
+            sigma = jnp.sqrt(dt_curr) * g
+            y_next = mu + sigma * noise_curr
+            return y_next, y_next
+
+        noises = random.normal(
+            key, shape=_sample_shape + (self.event_shape[0] - 1,) + self.event_shape[1:]
+        )
+        inits = self.init_dist.sample(key, sample_shape=_sample_shape)
+        scan_fn = lambda init, noise: scan(step, init, (noise, self.t[:-1], self.dt))
+
+        if sample_shape:
+            _, sde_out = vmap(scan_fn)(inits, noises)
```

Comment:
How does it work ?
Should t and dt also have sample shape ?

Reply:
Do you want `t` to have batch dimensions? If you think it is a good idea to have batched t, then you'll need to
+ broadcast `t` and `dt` to `sample_shape + self.batch_shape + dt.shape[-1:]`
+ reshape them to `(-1, dt.shape[-1:])`
+ vmap over inits, noises, t, dt

It would be nice (but seems unnecessary) to support batched `t` though.

Reply:
> Do you want t to have batch dimensions?
No. I assume the shapes as below.
`inits.shape = sample_shape_ + self.event_shape[1:] `
`noises.shape = sample_shape_ + (self.event_shape[0] - 1,) + self.event_shape[1:]`
`self.t[:-1].shape = (self.event_shape[0] - 1,)`
`self.dt.shape = (self.event_shape[0] - 1 ,)`

In this case, I don't know how I can apply vmap(scan)) to them.